### PR TITLE
Add support for LoRA with Transformer Engine

### PR DESCRIPTION
**Pull request by balvisio**
This PR adds support so that TransformerEngine layers (https://github.com/NVIDIA/TransformerEngine) are recognized as valid layers to which LoRA adapters can be added.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Thanks for this PR to add support for TE in PEFT. Before proceeding further, do you have a practical example of how to use it with a transformers model? I assume one way would be for the user to employ accelerate to apply TE. Also, I saw that you intend to replace the `nn.Linear` layers used for `lora_A` and `lora_B` with TE's low FP precision layers. Did you test if that works well? Usually, we would keep these layers in fp32 (or fp16/bf16), even if the base layer uses lower precision (say, int4).

**Comment by balvisio** [comment]
Thank you for taking a look at this. The primary goal of this PR is to support models with TE layers already inside them. In fact, using te.Linear for the adapters is not strictly necessary. However, by default, this TE layers will use the torch default type.
For the example, we have created and validated a "recipe" to use fine-tune an ESM2 model using LoRA here: https://github.com/NVIDIA/bionemo-framework/tree/main/bionemo-recipes/recipes/esm2_peft_te

**Comment by BenjaminBossan** [comment]
> The primary goal of this PR is to support models with TE layers already inside them.

I see, could you please provide a small example of how this looks in practice? We should eventually add a unit test with a real model architecture anyway, so this example would also help for that.

> using te.Linear for the adapters is not strictly necessary.

I think for consistency, we should stick with float32 by default. We could think of an option to use 4bit and 8bit floats but I'd have to think about how the API for that should look.

> For the example, we have created and validated a "recipe" to use fine-tune an ESM2 model using LoRA here: https://github.com/NVIDIA/bionemo-framework/tree/main/bionemo-recipes/recipes/esm2_peft_te

I think it would be great to include such an example here. Maybe simplified if possible (using `Trainer`, no DDP, no extra logging). WDYT?

**Comment by balvisio** [comment]
>  We should eventually add a unit test with a real model architecture anyway, so this example would also help for that.
There is a unit test that creates a toy model that uses TE layers. Is that what you meant ? Otherwise I can change the test to do: `AutoModelForTokenClassification.from_pretrained("nvidia/esm2_t6_8M_UR50D", config=config, trust_remote_code=True, dtype="bfloat16")`. lmk

> I think for consistency, we should stick with float32 by default.
Do you mean to remove using TE layers as adapters? I can do that but just to make sure, currently the TE adapters will use the default torch type, not a low a precision necessarily.

> I think it would be great to include such an example here
Where should I add the example exactly? Is there a doc with examples?

**Comment by BenjaminBossan** [comment]
> There is a unit test that creates a toy model that uses TE layers. Is that what you meant ? Otherwise I can change the test to do

Yeah, we can use a bit heavier models in `test_gpu_examples.py`, my goal would be to use a real model there, or at least a "tiny" version of a real model, instead of a toy model.

> Do you mean to remove using TE layers as adapters?

Yes, so keep the base layers as they are and use normal `nn.Linear` for LoRA.

> I can do that but just to make sure, currently the TE adapters will use the default torch type, not a low a precision necessarily.

Is there any advantage then?

> Where should I add the example exactly? Is there a doc with examples?

Docs would be nice, but I was thinking of the `examples/` folder.

**Comment by balvisio** [comment]
@BenjaminBossan : I changed the tests to use a TE-based model and added an example.

File: tests/test_lora_te.py
Comment (BenjaminBossan) [inline tests/test_lora_te.py:1]: Thanks for adding tests with a small but not toy model. Let's move the tests to `test_gpu_examples.py` and group them all into one class `TestTransformerEngine` (check the other tests there for appropriate decorators).

Unfortunately, it seems like these tests require a model with `trust_remote_code=True`, which we really really want to avoid running automatically. Do you know if there is a model that can be used with TE and doesn't require remote code execution (say, one of the ESM models directly provided by Transformers)?

File: tests/test_lora_te.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+def test_te_lora_backward(esm2_model, tokenized_inputs):
+    cfg = LoraConfig(target_modules=["layernorm_qkv"], r=4, lora_alpha=8)
+    lora_model = get_peft_model(esm2_model, cfg).to("cuda")
+
+    optimizer = torch.optim.AdamW(lora_model.parameters())
+    loss_fn = torch.nn.CrossEntropyLoss()
+    inputs = {k: v.to(next(lora_model.parameters()).device) for k, v in tokenized_inputs.items()}
+    logits = lora_model(**inputs).logits
+    labels = torch.randint(logits.shape[-1], logits.shape[:-1], device=logits.device)
+    loss = loss_fn(logits.view(-1, logits.size(-1)), labels.view(-1))
```
Comment (BenjaminBossan) [inline tests/test_lora_te.py]: How about also checking that the loss is finite or within a certain range?

File: src/peft/tuners/lora/te.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+from peft.tuners.tuners_utils import BaseTunerLayer
+
+from .config import LoraConfig
+
+
+if is_te_available():
+    import transformer_engine as te
+
+
+class TELoRA(torch.nn.Module, LoraLayer):
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/te.py]: This is a TE layer for linear layers, right? So how about naming it `TeLinear`.

File: examples/lora_finetuning_transformer_engine/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+torch
+transformers
+datasets
+peft
+accelerate
+numpy
+safetensors
+# transformer-engine must match the system CUDA toolkit version.
+# Install it separately or use the system-provided package.
+# See: https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/installation.html
```
Comment (BenjaminBossan) [inline examples/lora_finetuning_transformer_engine/requirements.txt:12]: For other folks who want to test, I had to manually set the `CPLUS_INCLUDE_PATH` env var to point to `cudnn` and `nccl`.

File: examples/lora_finetuning_transformer_engine/lora_finetuning_te.py
Comment (BenjaminBossan) [inline examples/lora_finetuning_transformer_engine/lora_finetuning_te.py:1]: This script runs code with `trust_remote_code=True`. As a user, I would be quite unhappy about that. Let's make this a command line argument that defaults to `False`, requiring users to opt-in. Of course, it would mean that there is an error when running the script with the default model. So I would suggest to make a check at the start if `trust_remote_code is False` and ESM2 is being used to raise an error and tell the user they need to enable `trust_remote_code` for this model.

File: src/peft/tuners/lora/te.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                lora_B = self.lora_B[active_adapter]
+                dropout = self.lora_dropout[active_adapter]
+                scaling = self.scaling[active_adapter]
+                x = self._cast_input_dtype(x, lora_A.weight.dtype)
+                result = result + lora_B(lora_A(dropout(x))) * scaling
+
+            result = result.to(torch_result_dtype)
+
+        return result
+
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/te.py:106]: Would merging/unmerging work with this layer?

File: examples/lora_finetuning_transformer_engine/README.md
Comment (BenjaminBossan) [inline examples/lora_finetuning_transformer_engine/README.md:1]: I would suggest to rename the directory to include "transformer_engine" in the name, many users may not know what "te" means.

File: examples/lora_finetuning_transformer_engine/requirements.txt (hunk 3 above)
Comment (balvisio) [reply examples/lora_finetuning_transformer_engine/requirements.txt:12]: To test this i used the the NVIDIA pytorch image ` nvcr.io/nvidia/pytorch:26.01-py3` which is publicly available. I can add instructions in the README. sounds good?

File: tests/test_lora_te.py
Comment (balvisio) [reply tests/test_lora_te.py:1]: Unfortunately, ATM there are no TE based models in the Transformers library.

File: tests/test_lora_te.py (hunk 1 above)
Comment (balvisio) [reply tests/test_lora_te.py]: done

File: examples/lora_finetuning_transformer_engine/README.md
Comment (balvisio) [reply examples/lora_finetuning_transformer_engine/README.md:1]: done

File: examples/lora_finetuning_transformer_engine/lora_finetuning_te.py
Comment (balvisio) [reply examples/lora_finetuning_transformer_engine/lora_finetuning_te.py:1]: done

File: src/peft/tuners/lora/te.py (hunk 4 above)
Comment (balvisio) [reply src/peft/tuners/lora/te.py:106]: good point. It wouldn't right now. It would require some thought in case of `fp8`. If it is ok , I can create an issue for now and I can work on the functionality when needed.

File: examples/lora_finetuning_transformer_engine/README.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
+# Transformer Engine ESM2 LoRA Fine-Tuning
+
+This example demonstrates LoRA fine-tuning for Transformer Engine ESM2 token classification. It uses a synthetic dataset generated at runtime.
```
Comment (BenjaminBossan) [inline examples/lora_finetuning_transformer_engine/README.md]: Is there a real dataset on HF that could be used instead? Checking your original repo, it looks like you're using files from disk there, so if there is no good dataset on HF, synthetic is fine as well.

File: examples/lora_finetuning_transformer_engine/requirements.txt (hunk 3 above)
Comment (BenjaminBossan) [reply examples/lora_finetuning_transformer_engine/requirements.txt:12]: Yes, why not, it could be helpful for some users.

File: src/peft/tuners/lora/te.py (hunk 4 above)
Comment (BenjaminBossan) [reply src/peft/tuners/lora/te.py:106]: Let's just raise a `NotImplementedError` for `merge` and `unmerge` with a nice error message, and document that those are not supported yet.

File: src/peft/tuners/lora/te.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: LicenseRef-Apache2
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/te.py:2]: I haven't seen "LicenseRef-Apache2" before so I asked ChatGPT and it said that the correct way of specifying this SPDX license would be "SPDX-License-Identifier: Apache-2.0", providing [this reference](https://spdx.dev/learn/handling-license-info/). I'm not an expert at all, just wanted to mention it.

File: tests/test_gpu_examples.py [hunk 7]
```diff
@@ -5838,3 +5840,89 @@ def test_lora_8bit_bnb_dtype_autocast_float32(self, peft_type, tmp_path):
...
+
+@pytest.mark.skipif(not is_te_available(), reason="transformer_engine is not available")
+class TestTransformerEngine:
+    """Tests for LoRA with TransformerEngine layers."""
+
+    model_id = "nvidia/esm2_t6_8M_UR50D"
+
+    @pytest.fixture
+    def config(self):
+        return AutoConfig.from_pretrained(self.model_id, trust_remote_code=True)
```
Comment (BenjaminBossan) [inline tests/test_gpu_examples.py]: We can't use `trust_remote_code=True` in our tests for external repos. So unfortunately, we need to figure out a different solution here. Would it be possible without too much extra effort to take another small model, say `facebook/opt-125m`, and replace some its layers with TE layers?

Comment (balvisio) [reply tests/test_gpu_examples.py]: done

File: src/peft/tuners/lora/te.py (hunk 6 above)
Comment (balvisio) [reply src/peft/tuners/lora/te.py:2]: ok. let me check internally but should be fine since is what we use for other repos too.

File: src/peft/tuners/lora/te.py (hunk 4 above)
Comment (balvisio) [reply src/peft/tuners/lora/te.py:106]: done

File: examples/lora_finetuning_transformer_engine/README.md (hunk 5 above)
Comment (balvisio) [reply examples/lora_finetuning_transformer_engine/README.md]: We are working on the licensing to be able to upload the data to HF. It is a bit tricky because it is bio related data. For now I added a pointer to a script that the users can run to download it.

**Comment by balvisio** [comment]
@BenjaminBossan Thanks for looking at this. I have addressed your comments.

**Comment by BenjaminBossan** [comment]
@balvisio Thanks for the updates, the PR looks good. Could you please run `make style`?

**Comment by balvisio** [comment]
done
